### ISSUE TITLE
Sometimes I get a `google.auth.exceptions.RefreshError`

### ISSUE THREAD

**Issue by doekman** [googleapis/google-auth-library-python#1142]
We use oauth2 for our internal web application. However, one user gets an RefreshError at a certain point in the application. The user can log-in, but the first call to drive_services.files().list(...) results in a `RefreshError` with the added information `('invalid_grant: Bad Request', {'error': 'invalid_grant', 'error_description': 'Bad Request'})`.

Exception:

	Traceback (most recent call last):
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/flask/app.py", line 1523, in full_dispatch_request
	    rv = self.dispatch_request()
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/flask/app.py", line 1509, in dispatch_request
	    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
	  File "/home/webby/aiwa/aiwa/zaken.py", line 93, in project_detail
	    folder_id, nr_folders_created = mkdir_p(drive_service, info['shared_drive_id'], map_naam, parent_id=parent_folder_id)
	  File "/home/webby/aiwa/aiwa/adapters/g00gle.py", line 341, in mkdir_p
	    result = list(search_files(drive_service, drive_id, parent_id=parent_id, mime_type='application/vnd.google-apps.folder', filename=dir_name))
	  File "/home/webby/aiwa/aiwa/adapters/g00gle.py", line 242, in search_files
	    response = drive_service.files().list(corpora="drive",
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
	    return wrapped(*args, **kwargs)
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/googleapiclient/http.py", line 923, in execute
	    resp, content = _retry_request(
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/googleapiclient/http.py", line 191, in _retry_request
	    resp, content = http.request(uri, method, *args, **kwargs)
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/google_auth_httplib2.py", line 245, in request
	    self.credentials.refresh(self._request)
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/google/oauth2/credentials.py", line 302, in refresh
	    ) = reauth.refresh_grant(
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/google/oauth2/reauth.py", line 347, in refresh_grant
	    _client._handle_error_response(response_data)
	  File "/home/webby/.cache/pypoetry/virtualenvs/aiwa-2k8Db8lp-py3.9/lib/python3.9/site-packages/google/oauth2/_client.py", line 62, in _handle_error_response
	    raise exceptions.RefreshError(error_details, response_data)
	google.auth.exceptions.RefreshError: ('invalid_grant: Bad Request', {'error': 'invalid_grant', 'error_description': 'Bad Request'})

As far as I can tell is my code correct. Other users have no problems. What can I do?

**Comment by clundin25** [comment]
If I understand correctly this issue is only happening to one particular user?

To clarify, do you mean only the first call to `drive_services.files().list(...)` has this issue and subsequent calls to `drive_services.files().list(...)` pass? Or did you mean that the first call will halt all progress?

Can the user try to provision a new Refresh token? It may have issues that prevent it from being used to acquire a new access token.

**Comment by doekman** [comment]
At the moment, this bug only happens to one particular user. But not all users have permission to use functionality that uses the drive-API.

And yes, the first call to `drive_services.files().list(...)` fails with a RefreshError. There are no subsequent calls. So yes, the first call halts all progress at the moment.

How would the user provision a new Refresh token? I'm not familiar with this.

**Comment by clundin25** [comment]
What flow are you using to authorize users? Having them re-authorize your application will generate a new refresh token. 

For example, `gcloud auth application-default login` can be used by developers to create a new refresh token.

**Comment by doekman** [comment]
I use [flask-dance](https://flask-dance.readthedocs.io/en/latest/providers.html#module-flask_dance.contrib.google) for the flow. I don't see an option to re-authorize, but maybe `reprompt_consent` can be used?

**Comment by clundin25** [comment]
I'm not very familiar with `flask-dance` but I will take a look later today.

Some reasons that the refresh token could be rejected can be found on this page https://developers.google.com/identity/protocols/oauth2

**Comment by clundin25** [comment]
From what I can tell this seems to be an issue with how the user is acquiring the refresh token. How long has this been a problem? How many times have they tried to login? How many times have they seen the screen to authorize your application to use the token provided by Google? They should get a new refresh token when going through the oauth screen.

Unfortunately I do not see anything in these logs or details that indicate a problem with the client library. If you still have issues I suggest opening a support case and the that team may be able to get visibility on what's happening on the server side.

**Comment by doekman** [comment]
I might have found a lead. I modelled the account creation according to [this example](https://flask-dance.readthedocs.io/en/latest/multi-user.html#overriding-the-default-behavior). When it's a new local account, the token is stored in a new database record. However, when the local account already exists, the new token is not stored (the if `if oauth.user:` in the middle).

When only the access-token has been updated, this is not a problem. The Google Drive API will request a new access-token by itself. However, when the "dance" gets a handed out new refresh-token, it will not be stored in the example. 

At least, that's my hypothesis. I'll test tomorrow.
